### ISSUE TITLE
timing of tasks.

### ISSUE THREAD

**Issue by dh-metre** [WarriorOfWire/CircuitPython_async#1]
ISSUE:  At above  16Hz the system freezes.  If i set the task frequency above 8hz the measured freq gets pegged to 16hz no matter what the setting.   However the iteration count seems to match the report interval for the task that is doing the reporting.  i put some time.monotonic_ns timing around all of these. The tasks are all completing in under 0.5 ms on average. 

Background:

Thank you for this interesting and helpful library i have been building an application around it.

Things have generally been working as described in the documentation but at some point when my application grew to have multiple modules and objects i started to see the scheduler freeze.

i have an event loop and usb, hardware uart and ble uart buffer parsing and and event loop for a total of 4x tasks. orginally the  slower usb, ble and event tasks were at 10hz and the uart was at 100hz, because of a lot of serial data streaming.  I had all of this working great balanced the buffers etc.

**Comment by dh-metre** [comment]
Here is my repro. I turned on logging in tasko.  It appears that the float returned via time.monotonic() does not have enough resolution.  It evaluates as zero and appears to enter an infinite loop.

```
from metre_lib import tasko
import time

async def task1():
    print("task1: ", time.monotonic())

async def task2():
    print("task2: ", time.monotonic())

async def task3():
    print("task3: ", time.monotonic())

tasko.schedule(hz=10, coroutine_function=task1)
tasko.schedule(hz=15, coroutine_function=task2)
tasko.schedule(hz=20, coroutine_function=task3)

tasko.run()
```

```
Press any key to enter the REPL. Use CTRL-D to reload.
Adafruit CircuitPython 6.0.1 on 2020-12-28; Adafruit ItsyBitsy nRF52840 Express with nRF52840
>>> 
>>> import tasko_test
Enabling tasko instrumentation
Enabling tasko instrumentation
scheduling  <generator>  at  10 hz
adding task  <coroutine object 'schedule_at_rate' at 20014d40>
scheduling  <generator>  at  15 hz
adding task  <coroutine object 'schedule_at_rate' at 20014e20>
scheduling  <generator>  at  20 hz
adding task  <coroutine object 'schedule_at_rate' at 20014f30>
stepping over  3  tasks
hz:   10.00000
seconds_per_invocation:    0.10000
target_run_time1: 270019.50741
iteration  <coroutine object 'task1' at 200150b0>
task1:  270020.0
target_run_time2: 270019.50741
target_run_time3: 270019.50741
seconds_to_wait:    0.00000
iteration  <coroutine object 'task1' at 20015140>
task1:  270020.0
target_run_time2: 270019.50741
target_run_time3: 270019.50741
seconds_to_wait:    0.00000
```

**Comment by dh-metre** [comment]
FWIW:  Refactored loop.py replacing time.monotonic() with time.monotonic_ns().
[loop.py.zip](https://github.com/WarriorOfWire/tasko/files/6045224/loop.py.zip)

**Comment by kvc0** [comment]
Oh hey, I saw someone link this issue in discord but GitHub never notified me! I’ll take a look at this soon, or if you have a fix feel free to open a PR. I’ll fix my notification settings.

**Comment by kvc0** [comment]
@dh-metre I've updated with some scheduling improvements. I could not repro the issue you mentioned but I've added more safety around sleep and have added a performance baseline test.

Scheduling at high frequency is a normal thing to do with this library. I routinely use it at 100hz+ with polling libraries on m4's. I haven't spent much time with nrf52840's but I've got one on my desk to try out sometime!
